zstd/_generate/gen.go

```diff
@@ -287,12 +287,51 @@ func (o options) generateBody(name string, executeSingleTriple func(ctx *execute
 
 	// Update states, max tablelog 28
 	{
-		Comment("Update Literal Length State")
-		o.updateState(name+"_llState", llState, brValue, brBitsRead, "llTable")
-		Comment("Update Match Length State")
-		o.updateState(name+"_mlState", mlState, brValue, brBitsRead, "mlTable")
-		Comment("Update Offset State")
-		o.updateState(name+"_ofState", ofState, brValue, brBitsRead, "ofTable")
+		if o.bmi2 {
+			// Get total number of bits (it is safe, as nBits is <= 9, thus 3*9 < 255)
+			total := GP64()
+			LEAQ(Mem{Base: llState, Index: mlState, Scale: 1}, total)
+			ADDQ(ofState, total)
+			MOVBQZX(total.As8(), total) // total = llState.As8() + mlState.As8() + ofState.As8()
+
+			// Read `total` bits
+			bits := o.getBitsValue(name+"_getBits", total, brValue, brBitsRead)
+
+			// Update states
+			Comment("Update Offset State")
+			{
+				nBits := GP64()
+				lowBits := GP64()
+				MOVBQZX(ofState.As8(), nBits)
+				BZHIQ(nBits, bits, lowBits) // lowBits = bits & ((1 << nBits) - 1))
+				SHRXQ(nBits, bits, bits)    // bits >= nBits
+				o.nextState(name+"_ofState", ofState, lowBits, "ofTable")
+			}
+			Comment("Update Match Length State")
+			{
+				nBits := GP64()
+				lowBits := GP64()
+				MOVBQZX(mlState.As8(), nBits)
+				BZHIQ(nBits, bits, lowBits) // lowBits = lowBits & ((1 << nBits) - 1))
+				SHRXQ(nBits, bits, bits)    // lowBits >= nBits
```

Comment:
SHRQ will do `&63`, so you can omit `MOVBQZX` above. Not sure if SHRXQ does, but you can try.
```suggestion
				SHRQ(nBits, bits, bits)    // lowBits >>= nBits
```

Reply:
oh, yeah, you must use CL, unless SHRXQ works.

Reply:
SHRXQ also does `&63`